dev/src/overmind/state.tsx

```diff
@@ -187,7 +207,96 @@ export const state: State = {
     courses: [],
     courseSubmissions: [],
     courseGroupSubmissions: {},
+    sortedAndFilteredSubmissions: derived((state: State, rootState: Context["state"]) => {
+        // Filter and sort submissions based on the current state
+        if (!state.activeCourse || !state.courseSubmissions[state.activeCourse]) {
+            return []
+        }
+        let submissions = state.groupView
+            ? state.courseGroupSubmissions[state.activeCourse]
+            : state.courseSubmissions[state.activeCourse]
+
+        if (!submissions) {
+            return []
+        }
 
+        // If a specific assignment is selected, filter by that assignment
+        const numAssignments = rootState.review.assignmentID > 0
+            ? 1
+            : state.assignments[state.activeCourse].length ?? 0
+
+        let filteredSubmissions = submissions
+        for (const filter of state.submissionFilters) {
+            switch (filter) {
+                case "teachers":
+                    filteredSubmissions = filteredSubmissions.filter(submission => {
+                        return submission.enrollment ? !isTeacher(submission.enrollment) : false
+                    })
+                    break
+                case "approved":
+                    // approved filters all entries where all assignments have been approved
+                    filteredSubmissions = filteredSubmissions.filter(submission => {
+                        if (rootState.review.assignmentID > 0) {
+                            // If a specific assignment is selected, filter by that assignment
+                            const sub = submission.submissions?.find(submission => submission.getAssignment()?.getId() === rootState.review.assignmentID)?.getSubmission()
+                            return sub && isApproved(sub)
+                        }
+                        const numApproved = submission.submissions?.reduce((acc, cur) => {
+                            return acc + ((cur.hasSubmission() &&
+                                isApproved(cur.getSubmission() as Submission)) ? 1 : 0)
+                        }, 0) ?? 0
+                        return numApproved < numAssignments
+                    })
+                    break
+                default:
+                    break
+            }
+        }
+
+        const m = state.sortAscending ? -1 : 1
+        const sortedSubmissions = Object.values(filteredSubmissions).sort((a, b) => {
+            let subA: Submission | undefined
+            let subB: Submission | undefined
+            if (rootState.review.assignmentID > 0) {
+                // If a specific assignment is selected, sort by that assignment
+                subA = a.submissions?.find(submission => submission.getAssignment()?.getId() === rootState.review.assignmentID)?.getSubmission()
```

Comment:
These are similar to Line 241. Would it make sense to use a helper or something for this?